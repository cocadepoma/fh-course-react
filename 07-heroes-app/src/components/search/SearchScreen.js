import React, { useMemo } from "react";
// Libreria para obtener los params de la url de forma sencilla
// queryString.parse(location.search); Muestra todos los params que hay en la url
import queryString from "query-string";
import { useLocation } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { HeroeCard } from "../heroes/HeroeCard";
import { getHeroesByName } from "../../selectors/getHeroesByName";

export const SearchScreen = ({ history }) => {
    const location = useLocation();
    const { q = "" } = queryString.parse(location.search);

    const [formValues, handleInputChange] = useForm({ searchText: q });
    const { searchText } = formValues;

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);
    };

    return (
        <div>
            <h1>SearchScreen</h1>
            <hr />
            <div className='row'>
                <div className='col-5'>
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input
                            type='text'
                            name='searchText'
                            className='form-control'
                            placeholder='Find your hero'
                            value={searchText}
                            onChange={handleInputChange}
                            autoComplete='off'
                        />
                        <button type='submit' className='btn btn-outline-primary mt-3 w-100'>
                            Search...
                        </button>
                    </form>
                </div>
                <div className='col-7'>
                    <h4>Results</h4>
                    <hr />

                    {q === "" && <p className='alert alert-info'>Search a Hero</p>}
                    {q !== "" && heroesFiltered.length === 0 && (
                        <p className='alert alert-danger'>
                            The hero <b>{q}</b> doesn't exists
                        </p>
                    )}

                    {heroesFiltered.map((hero) => (
                        <HeroeCard key={hero.id} {...hero} />
                    ))}
                </div>
            </div>
        </div>
    );
};
