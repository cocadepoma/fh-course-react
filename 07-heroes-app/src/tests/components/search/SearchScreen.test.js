import React from "react";
import { mount } from "enzyme";
import { SearchScreen } from "../../../components/search/SearchScreen";
import { MemoryRouter, Route } from "react-router-dom";

describe("pruebas en <SearchScreen />", () => {
    test("debe de mostrarse correctamente con valores por defecto", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe("Search a Hero");
    });
    test("debe de recoger el param por la URL y mostrarlo en el input", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=green"]}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find("input").prop("value")).toBe("green");
        expect(wrapper.find("HeroeCard").exists()).toBe(true);
    });

    test("debe de mostrar un error si no se muestra el hero", () => {
        const q = "greendwdwd";
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${q}`]}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find(".alert-danger").text().trim()).toBe(`The hero ${q} doesn't exists`);
        expect(wrapper).toMatchSnapshot();
    });
    test("debe de llamar el push del history", () => {
        const q = "green";
        const history = {
            push: jest.fn(),
        };
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${q}`]}>
                <Route path='/search' component={() => <SearchScreen history={history} />} />
            </MemoryRouter>
        );
        wrapper.find("input").simulate("change", {
            target: {
                name: "searchText",
                value: q,
            },
        });
        wrapper.find("form").prop("onSubmit")({ preventDefault() {} });
        expect(history.push).toBeCalledWith(`?q=${q}`);
    });
});
