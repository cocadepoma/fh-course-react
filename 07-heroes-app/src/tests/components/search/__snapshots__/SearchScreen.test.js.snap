// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <SearchScreen /> debe de mostrar un error si no se muestra el hero 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="greendwdwd"
        />
        <button
          className="btn btn-outline-primary mt-3 w-100"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <p
        className="alert alert-danger"
      >
        The hero 
        <b>
          greendwdwd
        </b>
         doesn't exists
      </p>
    </div>
  </div>
</div>
`;

exports[`pruebas en <SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-3 w-100"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <p
        className="alert alert-info"
      >
        Search a Hero
      </p>
    </div>
  </div>
</div>
`;

exports[`pruebas en <SearchScreen /> debe de recoger el param por la URL y mostrarlo en el input 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="green"
        />
        <button
          className="btn btn-outline-primary mt-3 w-100"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "margin": "0 auto",
            "maxWidth": 340,
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Green Lantern"
              className="card-img"
              src="./assets/heroes/dc-green.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Green Lantern
              </h5>
              <p
                className="card-text"
              >
                Alan Scott, Hal Jordan, Guy Gardner, John Stewart, Kyle Raynor, Jade, Sinestro, Simon Baz
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  All-American Comics #16
                </small>
              </p>
              <a
                href="/hero/dc-green"
                onClick={[Function]}
              >
                Más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "margin": "0 auto",
            "maxWidth": 340,
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Green Arrow"
              className="card-img"
              src="./assets/heroes/dc-arrow.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Green Arrow
              </h5>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  More Fun Comics #73
                </small>
              </p>
              <a
                href="/hero/dc-arrow"
                onClick={[Function]}
              >
                Más...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
